### ISSUE TITLE
Custom Objective Behaving Badly (Or not training)

### ISSUE THREAD

**Issue by JoshuaC3** [dmlc/xgboost#3037]
Custom loss functions are behaving oddly. I tried to use the [custom fair loss function](https://www.kaggle.com/c/allstate-claims-severity/discussion/24520), which I have used successfully with XGBoost python (& Sklearn) before. However, today, for the first time, it is no longer working. The predictions are *all* 0.5 which means it is not training at all.

## Steps to reproduce

```
import xgboost as xgb

#make some easy to predict data
x = np.array(range(0, 100)).reshape(-1, 1)
y = range(0, 400, 4)

#define fair loss
def fair_obj(true, pred):
    x = true - pred
    c = 1
    den = abs(x) + c
    grad = (c*x) / den
    hess = (c*c) / (den ** 2)
    return grad, hess

xgbr = xgb.XGBRegressor(objective=fair_obj)

xgbr.fit(x, y)
print(xgbr.predict([1]))

[0.5]
```

Now try with a more simple loss function, mse,

```
def mse(true, pred):
    
    x = true - pred
    grad = x * 2
    hess = 2 * np.ones(x.shape)
    return grad, hess

xgbr = xgb.XGBRegressor(objective=mse)

xgbr.fit(x, y)
print(xgbr.predict([1]))

[-180824.36]
```

## What have you tried?

The following works (grad = **-** x * 2)

```
def mse(true, pred):
    
    x = true - pred
    grad = - x * 2
    hess = 2 * np.ones(x.shape)
    return grad, hess

xgbr = xgb.XGBRegressor(objective=mse)

xgbr.fit(x, y)
print(xgbr.predict([1]))

[4.094035]
```

But this negative gradient makes not sense to me. What is going on here? Moreover, how can I get the Fair Loss Function to work?

<details>
## Environment info
Operating System: Ubuntu

Compiler: python - pip installed
`xgboost` version used: xgb.__version__
'0.7.post3'

If you are using python package, please provide

Python version
3.6.3 |Anaconda custom (64-bit)| (default, Oct 13 2017, 12:02:49) 
[GCC 7.2.0]
Version info.
sys.version_info(major=3, minor=6, micro=3, releaselevel='final', serial=0)
</details>

**Comment by JoshuaC3** [comment]
Anything on this? Has anyone else had similar issues?
